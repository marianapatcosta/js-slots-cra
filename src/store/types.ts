
import { settingsReducer, slotMachineReducer } from './reducers';
import type { SettingsAction, SlotMachineAction } from './reducers';

export interface State {
  readonly settings: ReturnType<typeof settingsReducer>;
  readonly slotMachine: ReturnType<typeof slotMachineReducer>;
}

export type Action = SettingsAction | SlotMachineAction;