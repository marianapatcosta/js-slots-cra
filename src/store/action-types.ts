// SETTINGS
export const SOUND_STATE_CHANGED = 'soundStateChanged';
export const MUSIC_STATE_CHANGED = 'musicStateChanged';
export const THEME_CHANGED = 'themeChanged';
export const LANGUAGE_CHANGED = 'languageChanged';

// SLOT MACHINE
export const CREDITS_DECREASED = 'creditsDecreased';
export const CREDITS_INCREASED = 'creditsIncreased';
export const BET_UPDATED = 'betUpdated';
export const SPAN = 'span';
export const SPIN_ENDED = 'spinEnded';
export const AUTO_SPIN_STATE_CHANGED = 'autoSpinStateChanged';
export const SHOW_PAY_LINES_STATE_CHANGED = 'showPayLinesStateChanged';
export const GAME_RESET = 'gameReset';
export const RESET_MODAL_DISMISSED = 'resetModalDismissed';
export const GAME_LEFT = 'gameLeft';
export const NEW_SPIN_PREPARED = 'newSpinPrepared';
export const BONUS_WILD_CARDS_WON = 'bonusWildCardsWon';
