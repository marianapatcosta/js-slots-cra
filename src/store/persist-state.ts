import { Middleware } from 'redux';
import { COLLECTION_SAVED_STATE } from '@/constants';
import i18n from '@/locales/i18n';
import { setTheme } from '@/utils';
import { State } from './types';

const getStoredItem = <T>(collection: string): T | undefined => {
  const storedItem = localStorage.getItem(collection);
  if (!storedItem) {
    return;
  }
  const parsedStoredItem: T = JSON.parse(storedItem);
  return parsedStoredItem;
};

const saveItemInStorage = <T>(collection: string, item: T) => {
  const stringifiedItem = JSON.stringify(item);
  localStorage.setItem(collection, stringifiedItem);
};

export const saveState: Middleware<{}, State> = storeAPI => next => action => {
  const oldState = storeAPI.getState();
  const result = next(action);
  const newState = storeAPI.getState();

  if (oldState !== newState) {
    saveItemInStorage(COLLECTION_SAVED_STATE, newState);
  }

  return result;
};

export const loadState = () => {
  try {
    const persistedState: State | undefined = getStoredItem(COLLECTION_SAVED_STATE);
    if (!persistedState) {
      return undefined;
    }
    setTheme(persistedState.settings.currentTheme);
    i18n.changeLanguage(persistedState.settings.currentLanguage);
    return persistedState;
  } catch (error) {
    return undefined;
  }
};
