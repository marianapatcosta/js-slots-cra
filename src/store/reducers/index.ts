import { reducer as settingsReducer, Action as SettingsAction } from './settings';
import { reducer as slotMachineReducer, Action as SlotMachineAction } from './slot-machine';

export { settingsReducer, slotMachineReducer };
export type { SettingsAction, SlotMachineAction };
