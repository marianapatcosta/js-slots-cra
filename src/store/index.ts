import { applyMiddleware, compose, createStore, combineReducers } from 'redux';
import { loadState, saveState } from './persist-state';
import { settingsReducer, slotMachineReducer } from './reducers';

const persistedState = loadState();

const devtoolsCompose =
  typeof window === 'object' && (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? (window as any).__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({})
    : compose;

export const store = createStore(
  combineReducers({
    settings: settingsReducer,
    slotMachine: slotMachineReducer,
  }),
  persistedState,
  devtoolsCompose(applyMiddleware(saveState))
);
