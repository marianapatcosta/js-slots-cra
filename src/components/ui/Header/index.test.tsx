import { fireEvent, render, screen } from '@testing-library/react';
import { Header } from './index';
import { Provider } from 'react-redux';
import { store } from '@/store';

describe('Header', () => {
  it('renders without errors and matches snapshot', () => {
    const {
      container: { firstChild },
    } = render(
      <Provider store={store}>
        <Header />
      </Provider>
    );

    expect(firstChild).toMatchSnapshot();
  });

  it('executes `share` function on share button click', () => {
    render(
      <Provider store={store}>
        <Header />
      </Provider>
    );
    const mockShare = jest.fn().mockImplementation(() => Promise.resolve());
    Object.defineProperty(window, 'navigator', {
      value: {
        share: mockShare,
      },
    });
    const shareButton = screen.getByRole('button') as HTMLButtonElement;
    shareButton && fireEvent.click(shareButton);
    expect(shareButton).toBeTruthy();
    expect(mockShare).toHaveBeenCalled();
  });
});
