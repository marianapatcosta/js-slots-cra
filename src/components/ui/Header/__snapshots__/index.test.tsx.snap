// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header renders without errors and matches snapshot 1`] = `
<header
  class="header"
>
  <h1>
    Let's play the nerdiest slot machine on the earth!!!
  </h1>
  <button
    aria-label="Share JS Slots"
    class="button "
    title="Share JS Slots"
  >
    <svg
      viewBox="0 -22 512 511"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="m512 233.82-212.78-233.32v139.2h-45.238c-140.27 0-253.98 113.71-253.98 253.98v73.77l20.094-22.02c68.316-74.852 164.98-117.5 266.32-117.5h12.805v139.2z"
      />
    </svg>
  </button>
</header>
`;
